File: cc/codeforces/2.cc
// https://codeforces.com/contest/1187/problem/B
#include<bits/stdc++.h>
using namespace std;
long long int sz;
long long int n;
long long int solve(char s[], char name[]){
	// name = tolower(name);
	// std::transform(name.begin(), name.end(), name.begin(), ::tolower);
	long long int ans = -1;
	bool incl[strlen(name)] = {0};
	long long int it = 0;
	int flag;
	while(1){
		flag = 0;
		for (int i = 0; i < strlen(name); ++i)
		{
			if(!incl[i] && s[it] == name[i]){
				incl[i] = true;
				break;
			}
		}
		it++;

		for (int i = 0; i < strlen(name); ++i)
			if(incl[i])
				flag++;
		if(flag == strlen(name)) break;
		
	}
	return it;
}
int main(int argc, char const *argv[])
{
	cin>>sz;
	char s[sz];
	cin>>s;
	cin>>n;
	while(n--){
		char name[sz];
		cin>>name;
		cout<<solve(s, name)<<endl;
	}
	return 0;
}

File: cc/codeforces/2.py
def solve(name):
	ans = -1
	incl = [False for i in range(size)]
	it = 0
	i = 0	
	while it is not len(name):
		if s[i]	== name[it]:
			incl[i] = True
			it = it + 1
			if i > ans:
				ans = i
			i = 0

		i = i + 1	
	print(ans + 1)		

size = int(input())
s = input() # string
t = int(input())

while t != 0:
	name = input()
	solve(name)
	t = t - 1